include/boost/math/ccmath/next.hpp

```diff
@@ -0,0 +1,456 @@
+}
+
+template <typename T>
+constexpr T calc_min_shifted(const std::true_type&)
+{
+   return boost::math::ccmath::ldexp(tools::min_value<T>(), tools::digits<T>() + 1);
+}
+
+template <typename T>
+constexpr T calc_min_shifted(const std::false_type&)
+{
+   static_assert(std::numeric_limits<T>::is_specialized, "Type T must be specialized.");
+   static_assert(std::numeric_limits<T>::radix != 2, "Type T must be specialized.");
+
+   return boost::math::ccmath::scalbn(tools::min_value<T>(), std::numeric_limits<T>::digits + 1);
+}
+
+template <typename T>
+constexpr T get_min_shift_value()
+{
+   const T val = calc_min_shifted<T>(std::integral_constant<bool, !std::numeric_limits<T>::is_specialized || std::numeric_limits<T>::radix == 2>());
+   return val;
+}
+
+template <typename T, bool b = boost::math::tools::detail::has_backend_type_v<T>>
+struct exponent_type
+{
+    using type = int;
+};
+
+template <typename T>
+struct exponent_type<T, true>
+{
+    using type = typename T::backend_type::exponent_type;
+};
+
+template <typename T, bool b = boost::math::tools::detail::has_backend_type_v<T>>
+using exponent_type_t = typename exponent_type<T>::type;
+
+template <typename T>
+constexpr T float_next_imp(const T& val, const std::true_type&)
+{
+    using exponent_type = exponent_type_t<T>;
+    
+    exponent_type expon {};
+
+    int fpclass = boost::math::ccmath::fpclassify(val);
+
+    if (fpclass == FP_NAN)
+    {
+        return val;
+    }
+    else if (fpclass == FP_INFINITE)
+    {
+        return val;
+    }
+    else if (val <= -tools::max_value<T>())
+    {
+        return val;
+    }
+
+    if (val == 0)
+    {
+        return detail::get_smallest_value<T>();
+    }
+
+    if ((fpclass != FP_SUBNORMAL) && (fpclass != FP_ZERO) 
+        && (boost::math::ccmath::fabs(val) < detail::get_min_shift_value<T>()) 
+        && (val != -tools::min_value<T>()))
+    {
+        //
+        // Special case: if the value of the least significant bit is a denorm, and the result
+        // would not be a denorm, then shift the input, increment, and shift back.
+        // This avoids issues with the Intel SSE2 registers when the FTZ or DAZ flags are set.
+        //
+        return boost::math::ccmath::ldexp(boost::math::ccmath::detail::float_next(static_cast<T>(boost::math::ccmath::ldexp(val, 2 * tools::digits<T>()))), -2 * tools::digits<T>());
+    }
+
+    if (-0.5f == boost::math::ccmath::frexp(val, &expon))
+    {
+        --expon; // reduce exponent when val is a power of two, and negative.
+    }
+    T diff = boost::math::ccmath::ldexp(static_cast<T>(1), expon - tools::digits<T>());
+    if(diff == 0)
+    {
+        diff = detail::get_smallest_value<T>();
+    }
+    return val + diff;
+}
+
+//
+// Special version for some base other than 2:
+//
+template <typename T>
+constexpr T float_next_imp(const T& val, const std::false_type&)
```

Comment:
Do we actually need to support bases other than 2 in the constexpr case?  I would be inclined to remove this and static_assert that the radix is 2.